### PR TITLE
Support for Messenger Platform 1.1 features

### PR DESCRIPTION
**Pull request by eXeDK**

### CODE REVIEW

**Comment by DVassilev** [comment]
@remixz Can you merge this pleaes?

**Comment by adamjace** [comment]
please merge @remixz

**Comment by victorcotap** [comment]
Please merge @remixz

**Comment by Naramsim** [comment]
Please Merge

**Comment by Naramsim** [comment]
Would be useful to add: https://developers.facebook.com/docs/messenger-platform/thread-settings/greeting-text

To show an initial greeting :smile:

**Comment by iMicknl** [comment]
> Coverage decreased (-25.6%) to 74.39% when pulling 5bc3b89 on eXeDK:master into 1ac973a on remixz:master.

Why are there no tests included? I think it is quite important to add some tests, before this will be merged.

**Comment by eXeDK** [comment]
Bump

**Comment by remixz** [comment]
@eXeDK Hey! Thanks for your contribution 😄, sorry I'm so late. I'll agree with @iMicknl that tests should be included for 100% coverage, plus we'd probably want documentation as well. However, I'm gonna add you as a contributor to this repo now, since you've been super helpful in other issues, which I really appreciate. If you want to merge this just to get things started, and work on tests later, go for it. I'll make sure to keep an eye on things, and do a release once everything's good.

File: index.js [hunk 1]
```diff
@@ -104,6 +162,11 @@ class Bot extends EventEmitter {
           this._handleEvent('message', event)
         }
 
+        // handle echos
+        if (event.message && event.message.is_echo) {
```
Comment (charlesaraya) [inline index.js]: `if (event.message.is_echo) { ... }` would be enough.

File: index.js [hunk 2]
```diff
@@ -114,10 +177,24 @@ class Bot extends EventEmitter {
...
         }
+
+        // handle account_linking
+        if (event.account_linking && event.account_linking.status) {
+          if (event.account_linking.status == 'linked') {
+            this._handleEvent('accountLinked', event)
+          } else if (event.account_linking.status == 'unlinked') {
+            this._handleEvent('accountUnlinked', event)
+          }
+        }
```
Comment (charlesaraya) [inline index.js]: 1. Use the `===` and `!==` operators. The `==` and `!=` operators do type coercion and should not be used ([Code conventions for Javascript - Douglas Crockford](http://javascript.crockford.com/code.html))

2. Refactor:

```javascript
// handle account linking
if (event.account_linking.status === 'linked') {
    this._handleEvent('accountLinked', event)
}
if (event.account_linking.status === 'unlinked') {
    this._handleEvent('accountUnlinked', event)
}
```

**Comment by charlesaraya** [comment]
It should be `if ... else if ... else` instead of a whole bunch of `if`'s. It's a waste of execution time to keep checking falsy conditions, since calbacks types are muatually exclusive. For better performance use `switch` ([Writting Efficient Javascript](http://archive.oreilly.com/pub/a/server-administration/excerpts/even-faster-websites/writing-efficient-javascript.html)).

Plus, I would handle unknown events:
```javascript
if (event.oneCallbackType) {
    ... handle it ...
} else if (event.othercallbackType) {
    ... handle it ...
} else if {    // as many as needed
    ... handle these too ...
} else {
    console.error('Webhook received an unknown messaging event: ', JSON.stringify(event))
} 
```

File: index.js (hunk 2 above)
Comment (eXeDK) [inline index.js]: @charlesaraya you are welcome to do a pull request containing the specific changes to eXeDK:messenger-bot/master

File: index.js (hunk 1 above)
Comment (adamjace) [inline index.js]: not true. you would get a typeError if message was undefined

Comment (charlesaraya) [inline index.js]: Correct.
```js
if (event.message) {
    if (event.message.is_echo) {
         this._handleEvent('echoMessage', event)
    } else {
         this._handleEvent('message', event)
    }
}
```

File: index.js (hunk 2 above)
Comment (charlesaraya) [inline index.js]: To your repo then? ok

**Comment by remixz** [comment]
I went ahead and finished this PR with full test coverage and documentation. Thank you for your hard work on this! Released as 2.4.0.

**Pull request merged**